### PR TITLE
Pytests of the database

### PR DESCRIPTION
**Pull request by cclauss**

### CODE REVIEW

**Comment by samosaara** [comment]
I'm not really the unit testing expert here, but I wonder... Isn't there supposed to be a single valid output? If that is the case, can't we automate the test? Isn't the whole point of unit testing? Automation? Why are we not using a (unittest)[https://docs.python.org/3/library/unittest.html] framework?

**Comment by cclauss** [comment]
Last week's [comment about `unitest`](https://github.com/LazoCoder/Pokemon-Terminal/pull/60#issuecomment-310866236) pushed me to start building pytests.

The test automation is already in place at https://github.com/LazoCoder/Pokemon-Terminal/blob/master/.travis.yml#L16 and will happen on all pull requests once #76 is fixed.

I see many repos that have multiple different tests (tox, flake8, nose, hypothesis, coverage.py, pytest, etc.) and I see no reason why this repo can do the same.  Looking at [pytest's own automated testing](https://github.com/pytest-dev/pytest/blob/master/tox.ini) will make your eyes water.  We can steer clear of that level of complexity.  ;-)

**Comment by samosaara** [comment]
Actually, silly me I didn't realize pytest was actually a testing framework since you don't perform any assertions and I didn't heard of it before. It's ok, we can use pytest. So what I'm really discussing here is, shouldn't the tests be automated? Shouldn't we be doing assertions? What's the point of a non-automated unit testing?! It doesn't help, no one will read a bunch of logs through to find a bug.

**Comment by cclauss** [comment]
Sorry...  Perhaps I am not being clear.  Once #76 is complete, the output will look something like https://travis-ci.org/cclauss/flake8_for_pt/builds/249020190

**Comment by cclauss** [comment]
@samosaara Can you please look at these assertions and let me know if your results match mine?

```python
# print('From https://en.wikipedia.org/wiki/Pok%C3%A9mon#Generation_1 ...')
# counts = {'kanto': 151, 'johto': 100, 'hoenn': 135, 'sinnoh': 107, 'all': 493}
db = Database()
assert len(db.get_kanto()) == 151   # Passes
assert len(db.get_johto()) == 100   # Fails: assert 101 == 100
assert len(db.get_hoenn()) == 135   # Fails: assert 144 == 135
assert len(db.get_sinnoh()) == 107  # Fails: assert 121 == 107
assert db.get_extra()               # Fails: returns zero pokemon!'
```

**Comment by caedus75** [comment]
@cclauss The difference you're getting is caused by how `Database.__load_extra()` calls `Pokemon()`, based on the image's filename.

Currently, the `Extra` pokemon are assigned to a region, therefore, `Database.get_[region]` returns a bigger list than expected and `Database.get_extra()`, which selects pokemons with `region` set to `None`, returns an empty list.

**Comment by samosaara** [comment]
They don't. And it's my fault. I just noticed I broke the get extras method. Currently the method gets all pokemon without region from the main list, but in my type subcommand update I've did some naming tricks to get the types even from the extra pokemon. I'll fix it.

**Comment by cclauss** [comment]
@caedus75 Thanks for the correct steer...  I was lost in the woods about what was really going on in [`database.__load_extra`](https://github.com/LazoCoder/Pokemon-Terminal/blob/master/database.py#L231) so my tests are wrong.

@samosaara Please go ahead and fixes in a separate PR to master.  I will need some time to take my tests apart and put them back together again based on my new understanding.  I want to do development on tests separate from development on main functionality to steer clear of assumption bias and shared code between dev and test.

In general, here are a few things that would help:
* [ ] Better argument parsing #79 -- These are great ideas
* [ ] One unified dg.get_regions(region_name) instead of separate get_kanto(), get_johto(), etc.
    * This is leading to code bloat in single_argument_handler
    * flake8 rightly declares that single_ and multiple_argument_handler are too complex
* [ ] Get [unittest.py](https://github.com/LazoCoder/Pokemon-Terminal/blob/master/unittest.py) to function on its own for a separate set of test from pytest
    * This can be [turned on in .travis.yml](https://github.com/LazoCoder/Pokemon-Terminal/blob/master/.travis.yml#L17) when it can work without human intervention

**Comment by cclauss** [comment]
This PR is ready for review.  The only remaining failures are caused by:

```python
assert db.get_extra(), 'db.get_extra() returns no pokemon'
```

**Comment by cclauss** [comment]
Read for review.

**Comment by samosaara** [comment]
After rebasing master into your branch locally, the only error I'm getting is:
```
========================================================================== FAILURES ===========================================================================
_______________________________________________________________________ test_regions_two _______________________________________________________________________

    def test_regions_two():
        for region_name in region_dict:
            if tuple_store:
                _ts_test_region(region_name)
            else:
>               _test_region(region_name)

test_datbase.py:194:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

region_name = 'extra'

    def _test_region(region_name):
        db = Database()
        region_name = (region_name or 'extra').lower()
        # Database unfortunately makes db.__get_region() private :-(
        func = {
            "kanto": db.get_kanto,
            "johto": db.get_johto,
            "hoenn": db.get_hoenn,
            "sinnoh": db.get_sinnoh,
            "extra": db.get_extra
        }[region_name]
        pokemon_list = func()
        region_record = region_dict[region_name]
        start = region_record.start
        end = len(db) if region_name == "extra" else region_record.end
        # make sure there are no missing pokemon
>       assert len(pokemon_list) == end - start + 1
E       assert 24 == ((517 - 0) + 1)
E        +  where 24 = len([<database.Pokemon object at 0x7f4507a2e0b8>, <database.Pokemon object at 0x7f4507a2e278>, <database.Pokemon object at...ject at 0x7f
4507a37a58>, <database.Pokemon object at 0x7f4507a37908>, <database.Pokemon object at 0x7f4507a37b00>, ...])

test_datbase.py:168: AssertionError
============================================================= 1 failed, 23 passed in 0.39 seconds =============================================================
```
I managed to fix it here with:
image

File: .travis.yml [hunk 1]
```diff
@@ -2,19 +2,19 @@ sudo: false
...
 python:
-    - 3.5         # currently Travis trusty default Python3 is 3.5.2
-    #- 3.6.1
-    #- "nightly"  # currently points to 3.7-dev (3.7.0a0)
+    - "3.5.0"
+    - "3.6.1"
+    - "nightly"  # currently points to 3.7-dev
 install:
-    - pip install flake8 pytest
+    - pip install flake8, pytest
```
Comment (samosaara) [inline .travis.yml]: Wasn't this comma causing issues?

Comment (cclauss) [inline .travis.yml]: I just backed `.travis.yml` out of the PR...  There are only 5 files still in the PR and their filename are all of the form `test_*.py`.

**Comment by cclauss** [comment]
Thanks.  I also added complexity measures... https://travis-ci.org/LazoCoder/Pokemon-Terminal/builds/249996902#L457-L460

**Comment by cclauss** [comment]
This PR is ready for review.

**Comment by cclauss** [comment]
@LazoCoder Your thoughts on this one?  The test cases in these five files will help everyone to find and fix issues when making changes.

**Comment by LazoCoder** [comment]
Looks good

**Pull request merged**